src/app/clusters/commodity-metering-server/commodity-metering-server.cpp

```diff
@@ -0,0 +1,360 @@
+        return true;
+    }
+};
+} // namespace
+
+static void CleanUpIDs(DataModel::List<const uint32_t> & IDs)
+{
+    if (!IDs.empty() && IDs.data())
+    {
+        Platform::MemoryFree(const_cast<uint32_t *>(IDs.data()));
+        IDs = DataModel::List<const uint32_t>();
+    }
+}
+
+static void CleanupMeteredQuantityData(DataModel::List<Structs::MeteredQuantityStruct::Type> & aValue)
+{
+    if (aValue.data() != nullptr)
+    {
+        for (auto & item : aValue)
+        {
+            CleanUpIDs(item.tariffComponentIDs);
+        }
+
+        Platform::MemoryFree(aValue.data());
+        aValue = DataModel::List<Structs::MeteredQuantityStruct::Type>();
+    }
+}
+
+Instance::~Instance()
+{
+    if (!mMeteredQuantity.IsNull())
+    {
+        CleanupMeteredQuantityData(mMeteredQuantity.Value());
+    }
+    Shutdown();
+}
+
+CHIP_ERROR Instance::Init()
+{
+    mMeteredQuantity.SetNull();
+    mMeteredQuantityTimestamp.SetNull();
+    mMeasurementType.SetNull();
+    mMaximumMeteredQuantities.SetNull();
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+    return CHIP_NO_ERROR;
+}
+
+void Instance::Shutdown()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+static CHIP_ERROR CopyMeteredQuantityEntry(const Structs::MeteredQuantityStruct::Type & src,
+                                           Structs::MeteredQuantityStruct::Type & dest)
+{
+    dest.quantity = src.quantity;
+
+    if (!SpanCopier<uint32_t>::Copy(src.tariffComponentIDs, dest.tariffComponentIDs, kMaxTariffComponentIDsPerMeteredQuantityEntry))
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetMeteredQuantity(const DataModel::Nullable<DataModel::List<Structs::MeteredQuantityStruct::Type>> & newValue)
+{
+    assertChipStackLockedByCurrentThread();
+
+    if (NullableListEqual(newValue, mMeteredQuantity))
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (newValue.IsNull())
+    {
+        mMeteredQuantity.SetNull();
+    }
+    else
+    {
+        if (!mMeteredQuantity.IsNull())
+        {
+            CleanupMeteredQuantityData(mMeteredQuantity.Value());
+        }
+
+        const size_t len = newValue.Value().size();
+
+        if (len == 0)
+        {
+            mMeteredQuantity = MakeNullable(DataModel::List<Structs::MeteredQuantityStruct::Type>());
+        }
+        else if (len <= kMaxMeteredQuantityEntries)
+        {
+            Platform::ScopedMemoryBuffer<Structs::MeteredQuantityStruct::Type> buffer;
+
+            if (!buffer.Calloc(len))
+            {
+                return CHIP_ERROR_NO_MEMORY;
+            }
+
+            for (size_t idx = 0; idx < len; idx++)
+            {
+                // Deep copy each MeteredQuantityStruct in the newValue list
+                ReturnLogErrorOnFailure(CopyMeteredQuantityEntry(newValue.Value()[idx], buffer[idx]));
```

Comment:
If the copy for the first one succeeds and the copy for the second one fails, this leaks the TariffComponentIDs of the first one, no?

On failure you either have to CleanUpIDs on all the things where they have already been allocated, or you should have a more-automatic memory management setup here (e.g. a struct that inherits from MeteredQuantityStruct and has a pointer to its heap-allocated ids buffer, with a destructor that deletes the buffer, move/copy constructors that hand over or duplicate the buffer, etc, etc).